File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int validate(TreeNode* root,bool &balanced){
        if(root==nullptr)return 0;
        int left=validate(root->left,balanced);
        int right=validate(root->right,balanced);
        if(abs(left-right)>1)
            balanced=0;
        return 1+max(left,right);
    }
    bool isBalanced(TreeNode* root) {
        bool balanced=1;
        validate(root,balanced);
        return balanced;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        INF = 1e9

        def dfs(root):
            if not root:
                return 0
            left = dfs(root.left) + 1
            right = dfs(root.right) + 1
            if abs(left - right) <= 1:
                return max(left, right)
            return INF

        return True if dfs(root) < INF else False
